File: C11/백준/Gold/1967. 트리의 지름/트리의 지름.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Node {
    int to;
    int weight;
    struct Node* next;
} Node;

Node* tree[10001];
int visited[10001];
int farNode, maxDist;

void addEdge(int from, int to, int weight) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->to = to;
    newNode->weight = weight;
    newNode->next = tree[from];
    tree[from] = newNode;
}

void dfs(int node, int dist) {
    visited[node] = 1;
    if (dist > maxDist) {
        maxDist = dist;
        farNode = node;
    }
    for (Node* cur = tree[node]; cur != NULL; cur = cur->next) {
        if (!visited[cur->to]) {
            dfs(cur->to, dist + cur->weight);
        }
    }
}

int main() {
    int n;
    scanf("%d", &n);

    for (int i = 0; i <= n; i++) {
        tree[i] = NULL;
    }

    for (int i = 0; i < n - 1; i++) {
        int parent, child, weight;
        scanf("%d %d %d", &parent, &child, &weight);
        addEdge(parent, child, weight);
        addEdge(child, parent, weight);
    }

    if (n == 1) {
        printf("0\n");
        return 0;
    }

    memset(visited, 0, sizeof(visited));
    maxDist = 0;
    dfs(1, 0);

    memset(visited, 0, sizeof(visited));
    maxDist = 0;
    dfs(farNode, 0);

    printf("%d\n", maxDist);
    return 0;
}

File: Java/백준/Gold/1967. 트리의 지름/트리의 지름.java
import java.io.*;
import java.util.*;

class Main {
    static List<Node>[] tree;
    static boolean[] visited;
    static int farNode;
    static int maxDist;

    static class Node {
        int to, weight;
        Node(int to, int weight) {
            this.to = to;
            this.weight = weight;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        tree = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++) {
            tree[i] = new ArrayList<>();
        }

        for (int i = 0; i < n - 1; i++) {
            String[] input = br.readLine().split(" ");
            int parent = Integer.parseInt(input[0]);
            int child = Integer.parseInt(input[1]);
            int weight = Integer.parseInt(input[2]);
            tree[parent].add(new Node(child, weight));
            tree[child].add(new Node(parent, weight));
        }

        if (n == 1) {
            System.out.println(0);
            return;
        }

        visited = new boolean[n + 1];
        maxDist = 0;
        dfs(1, 0);

        visited = new boolean[n + 1];
        maxDist = 0;
        dfs(farNode, 0);

        System.out.println(maxDist);
    }

    static void dfs(int node, int dist) {
        visited[node] = true;
        if (dist > maxDist) {
            maxDist = dist;
            farNode = node;
        }
        for (Node next : tree[node]) {
            if (!visited[next.to]) {
                dfs(next.to, dist + next.weight);
            }
        }
    }
}